Used cosmos as db

lilyjma opened the pull request:
App is working but with following problems:

1. A logged in user can mark other user's tasks as complete or not and can delete other user's tasks
(it can't add tasks on behalf of other users though)

(**UPDATE 11/7**: this might've been a problem that existed in the original forked repo)
2. Tasks have correct owners when a new user creates an account and log in, but wrong owners if an existing user logs in; however, refreshing the page corrects the problem. (Still trying to figure out what went wrong as the front end is getting the right data.)

(**UPDATE 11/7**: new bug introduced)
3. A user making a new account with an existing email should trigger an error message saying there's an existing user with that email--the error message is sent to the browser but not displayed for some reason. 

After fixing these two issues, the next steps are: 

1. Use Key Vault to manage the key to Cosmos and other creds. (**UPDATE 11/7**:  DONE)
2. Use Identity as way to authorize log in 
3. Host the app using Azure App Service

Please feel free to suggest other services to use, other ways to doing things, and best practices in general for app building.

Commit:
working prototype with postgres

Commit:
this app uses cosmos instead of postgres

Commit:
working prototype with cosmos as db

Commit:
used key vault to store cosmos creds

lilyjma commented:
DONE
1. Use Key Vault to manage the key to Cosmos and other creds

jongio commented on app/models/models.py:
```diff
@@ -39,102 +60,144 @@ def hashed_password(password):
+
+    @staticmethod
+    def get_user_by_email(user_email):
+        query = "SELECT * FROM users u where u.email='" + user_email + "'"
```
Cosmos SQL is subject to SQL injection.  

Read this: https://en.wikipedia.org/wiki/SQL_injection and this https://michaelhowardsecure.blog/2019/03/05/cosmosdb-and-sql-injection/
And then re-write these queries as parameterized queries: https://azuresdkdocs.blob.core.windows.net/$web/python/azure-cosmos/4.0.0b5/index.html?highlight=parameterized%20queries#query-the-database

lilyjma replied:
Done.

jongio commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Flask + React + Postgres Starter 
+# Flask + React + Cosmos 
```
Seems like this would be a different sample.  So let's chat before you attempt to merge back into this repo.  We'll likely create a new one.

lilyjma replied:
Okay. I'll keep it as is for now and change after we chat.

jongio commented on static/src/index.js:
```diff
@@ -1,5 +1,7 @@
 import React from "react";
```
I suggest using https://create-react-app.dev/ to help setup the app, wire up dependencies, and so on.

lilyjma replied:
Hi Jon,

I'm not quite sure what you mean by this. My understanding is that since this is not  purely a React app so I can't really just follow the code structure given on that site. From what I've read there, I haven't found any material on integrating React with a framework like Flask; I only came across something on integration with an API backend. Upon Googling 'flask react boilerplate', one of the Github repos that came up was the the one that joshgav used to base his Flask-React-Postgres app on. 

Could you elaborate on what you mean here? I'm probably misunderstanding you or missing some pieces of info here. 

Thanks!

-Lily

jongio replied:
Create react app is kind of a standard these days for creating react apps, it handles a lot of the stuff that I'm seeing issues with above, like css/js dependency managements, etc.  It sets you up with a good base to work from.  You can put this on the backlog to do later if you'd like.

jongio commented on app/__init__.py:
```diff
@@ -1,15 +1,25 @@
+key = app.config["SECRET_KEY"]
+uri = app.config["COSMOS_DB_URI"]
+
+client = CosmosClient(uri, credential=key)
```
I'm not crazy about having the cosmos client as a global variable for the app.   I would like all the "cosmos" stuff to be encapsulated so the app is only coupled to that encapsulation versus directly to cosmos.

lilyjma replied:
Done, I think. Added a db.py in /utils and kept cosmos things there. Imported the db in models.py to use. I thought this way, if in the future, we want to use other db, we can keep all db things there and just import the appropriate dbs to use in models.py or other scripts.

Commit:
use dotenv to read env vars

Commit:
add dotenv to requirement.txt

Commit:
parameterized sql queries

Commit:
encapsulated all db things in utils/db.py

Commit:
pass in user/task as dict to create/upsert item

lilyjma commented:
More changes. Invited Maggie to be collaborator.

Commit:
updating readme

lilyjma committed:
Update README.md

Commit:
update readme

Commit:
Merge branch 'team_standup_app_cosmos' of https://github.com/lilyjma/Flask-React-Postgres into team_standup_app_cosmos

Commit:
fixed path to get env vars correctly

maggiepint commented on app/models/models.py:
```diff
@@ -39,102 +43,137 @@ def hashed_password(password):
+
     def __init__(self, task, user_id, status):
-        self.date = datetime.utcnow().date()
+        self.id = str(randint(_MIN, _MAX))
```
I would think Cosmos has a way to generate random identifiers for documents - what happens if you just *don't* assign an id? I would think it would auto generate one.

jongio replied:
Also, randint doesn't guarantee unique across all items, so please switch over to using Guid for ids.  If you don't provide an "id" field, then Cosmos should auto create an id field and insert a guid.

lilyjma replied:
You have to add an id to each item you insert into the container according to the docs here. 
https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/cosmos/azure-cosmos#insert-data

I'll try to use Guid.

Commit:
update readme

lilyjma committed:
Update README.md

lilyjma committed:
Update README.md

lilyjma committed:
Update README.md

jongio commented on README.md:
```diff
@@ -1,68 +1,119 @@
+## 1. Setting up Azure Services
+First, sign up for a free [Azure](https://azure.microsoft.com/en-us/free/) account if you don't already have one. Sign into https://portal.azure.com.
+
+[Create a resource group](https://github.com/lilyjma/azurethings/blob/master/createResourceGroup.md) to store the resources that you'll be using here--Azure Cosmos DB and Key vault. Then follow the instructions in the links below to create each resource:
```
This references your fork directly.  Make a relative path please.

lilyjma replied:
I thought this should be in a repository of its own, because I created a repository called 'azurethings' to store small tutorials/things that I learned about Azure.

jongio commented on README.md:
```diff
@@ -1,68 +1,119 @@
+   1. When creating the database, name it 'team_standup'. For this app, you also need two containers named 'tasks' and 'users'. The [partition key](https://docs.microsoft.com/en-us/azure/cosmos-db/partitioning-overview#choose-partitionkey) for both is '/id'. 
+2. [Key Vault](https://docs.microsoft.com/en-us/azure/key-vault/quick-create-portal#create-a-vault)
+   1. This will store the credentials for the resources this app uses. For example, it'll store the key to the cosmos database. This way, you don't reveal any key in your code. 
+   2. You'll add two secrets called 'cosmosKey' and 'cosmosURI' to Key Vault to hold the cosmos key and URI respectively. To find these, click into the Cosmos DB account created, go to 'Keys' tab and get the Primary Key and the URI. 
```
Anytime you use 'cosmos' it should be "Cosmos DB" to adhere to branding

jongio commented on README.md:
```diff
@@ -1,68 +1,119 @@
+
+(Remember to store them in the resource group you created; this will make it easier to clean up the resources in the future.)
+
+1. [Azure Cosmos DB](https://docs.microsoft.com/en-us/azure/cosmos-db/create-cosmosdb-resources-portal#create-an-azure-cosmos-db-account)
```
I like to include the CLI commands along side the portal instructures or a section which is 

## Azure Portal
1. Portal steps here
## Azure CLI
1. CLI steps here

lilyjma replied:
The CLI commands are also executed on Portal though--in the Cloud Shell. You could also execute CLI commands on your local machine but that requires some extra configuration so I thought it'd be the easiest to just do it on Cloud Shell.

jongio commented on README.md:
````diff
@@ -1,68 +1,119 @@
+    
+    ```az ad sp create-for-rbac --name http://my-applications-url --skip-assignment```
+
+    Use your web app's url. To find that, click on the app in Portal, go to the 'Overview' tab and look for 'URL' on the top right portion of the page. It looks something like https://myUniqueAppName.azurewebsites.net. 
````
What do you mean use your web apps url?  In the above or below?  Please be explicit in everything you ask the user to do.

jongio commented on README.md:
````diff
@@ -1,68 +1,119 @@
-cd flask-react-postgres
-```
+   ```bash
+   git clone https://github.com/lilyjma/Flask-React-Postgres/tree/team_standup_app_cosmos
````
I don't think this path is right, but doesn't matter for now

jongio commented on README.md:
````diff
@@ -1,68 +1,119 @@
+   pip install -r requirements.txt
+   ```
 
 3. Import the project folder into VS Code
````
I think you mean "Open", not "Import"

jongio commented on README.md:
````diff
@@ -1,68 +1,119 @@
+   code .
+   ```
 
-## 2. Running The Code Locally
````
Please containerize each part of the app so the user just has to start the container to run it.  Talk with me if you have questions.

jongio commented on README.md:
````diff
@@ -1,68 +1,119 @@
+   ```
 
-## 2. Running The Code Locally
+4. Create a .env file in the root directory
````
Typically we create a .env.tmp file in the root with the exact structure and keys required, without the secrets.  And then ask the user to rename it to .env and put in the correct values.

Make sure .env is in your .gitignore if you do this.

jongio commented on app/__init__.py:
```diff
@@ -1,15 +1,23 @@
+    APP_DIR, "../static/build"
+)  # Where your index.html file is located
+
+dotenv_path = Path(".") / ".env"
```
dotenv It looks for .env in the root by default, so you shouldn't need this line.

jongio commented on app/models/models.py:
```diff
@@ -39,102 +43,137 @@ def hashed_password(password):
+    def get_user_by_email(user_email):
+        users = user_container.query_items(
+            query="SELECT * FROM users u WHERE u.email = @userEmail",
+            parameters=[dict(name="@userEmail", value=user_email)],
```
just @email would be fine here.

jongio commented on app/models/models.py:
```diff
@@ -39,102 +43,137 @@ def hashed_password(password):
-        if user and bcrypt.check_password_hash(user.password, password):
-            return user
+        users = user_container.query_items(
+            query="SELECT * FROM users u WHERE u.email = @userEmail AND u.password = @userPassword",
```
Just @ email and @ password should be ok.

jongio commented on app/models/models.py:
```diff
@@ -39,102 +43,137 @@ def hashed_password(password):
             return True
-        except IntegrityError:
+        except exceptions.CosmosHttpResponseError:  # item couldn't be upserted
             return False
```
Let's do some research on how we want to handle error messages from the model. Typically the UI doesn't want to show the error details, but something to consider, versus just returning false.

lilyjma replied:
Oh the UI wouldn't just show 'False.' It'll have an error message that's appropriate for the specific situation. Please take a look at the routes.py script.

jongio commented on app/models/models.py:
```diff
@@ -39,102 +43,137 @@ def hashed_password(password):
         return user
 
     @staticmethod
     def get_user_with_email_and_password(email, password):
```
We should be using OAuth to manage users and permissions.  Having a method with email and password basically means that your app controls user access, permissions, etc and storing of credentials in the database.  Typically you avoid with this access tokens and OAuth.  It's a bigger more complicated implementation that you should learn, but we can discuss later.

lilyjma replied:
Yes, I'd love to have a discussion on this. Right now, one problem I'm seeing is that when a user is logged in, he/she can edit other people's tasks, instead of just his/her own. So basically, a logged in user has access to the whole database, instead of just part of it that stores his/her own info. I don't know if this can be solved by OAuth, but the way this works now is not ideal. 

Also, can we discuss the create-react-app too, like how to make it integrate with Flask? I didn't change the structure of the code at all. I think the original app is based on this [boilerplate](https://github.com/dternyak/React-Redux-Flask) that combines Flask and React, which I guess is not ideally structured.

Commit:
used uuid for task id

Commit:
update readme

Commit:
added .env.tmp

Commit:
update readme

Commit:
update readme: authorize service principal to perform operations to keyvault

lilyjma committed:
Update README.md

lilyjma committed:
Update README.md

lilyjma committed:
Service principal change

Changed what needs to be done to create a service principal. Previously misunderstood the --name parameter in the command--it just has to be a name that starts with 'http://' and doesn't have to be the URI to a real, existing site.
